include/directx/d3dx12.h

```diff
@@ -4037,6 +4038,277 @@ class CD3DX12_NODE_MASK_SUBOBJECT
+    FEATURE_SUPPORT_GET(UINT, m_dGPUVASupport, MaxGPUVirtualAddressBitsPerProcess);
+
+    // 7: Shader Model
+    D3D_SHADER_MODEL HighestShaderModel(D3D_SHADER_MODEL RequestedShaderModel = D3D_HIGHEST_SHADER_MODEL)
+    {
+        if (RequestedShaderModel <= m_dShaderModel.HighestShaderModel) {
+            return RequestedShaderModel;
+        }
+
+        D3D12_FEATURE_DATA_SHADER_MODEL dShaderModel;
+        m_pDevice->CheckFeatureSupport(D3D12_FEATURE_SHADER_MODEL, &dShaderModel, sizeof(D3D12_FEATURE_DATA_SHADER_MODEL));
+        return dShaderModel.HighestShaderModel;
+    }
+
+    // 8: D3D12 Options1
+    FEATURE_SUPPORT_GET(BOOL, m_dOptions1, WaveOps);
+    FEATURE_SUPPORT_GET(UINT, m_dOptions1, WaveLaneCountMin);
+    FEATURE_SUPPORT_GET(UINT, m_dOptions1, WaveLaneCountMax);
+    FEATURE_SUPPORT_GET(UINT, m_dOptions1, TotalLaneCount);
+    FEATURE_SUPPORT_GET(BOOL, m_dOptions1, ExpandedComputeResourceStates);
+    FEATURE_SUPPORT_GET(BOOL, m_dOptions1, Int64ShaderOps);
+
+    // 10: Protected Resource Session Support
+    D3D12_PROTECTED_RESOURCE_SESSION_SUPPORT_FLAGS ProtectedResourceSessionSupport(UINT NodeIndex = 0) const
+    {
+        return m_dProtectedResourceSessionSupport[NodeIndex].Support;
+    }
+
+    // 12: Root Signature
+    D3D_ROOT_SIGNATURE_VERSION HighestRootSignatureVersion(D3D_ROOT_SIGNATURE_VERSION RequestedVersion = D3D_HIGHEST_ROOT_SIGNATURE_VERSION) const
+    {
+        if (RequestedVersion <= m_dRootSignature.HighestVersion)
+        {
+            return RequestedVersion;
+        }
+
+        D3D12_FEATURE_DATA_ROOT_SIGNATURE dRootSignature;
+        dRootSignature.HighestVersion = RequestedVersion;
+        m_pDevice->CheckFeatureSupport(D3D12_FEATURE_ROOT_SIGNATURE, &dRootSignature, sizeof(D3D12_FEATURE_DATA_ROOT_SIGNATURE));
+        return dRootSignature.HighestVersion;
+    }
+
+    // 16: Architecture1
+    FEATURE_SUPPORT_GET_NODE_INDEXED(BOOL, m_dArchitecture1, TileBasedRenderer);
+    FEATURE_SUPPORT_GET_NODE_INDEXED(BOOL, m_dArchitecture1, UMA);
+    FEATURE_SUPPORT_GET_NODE_INDEXED(BOOL, m_dArchitecture1, CacheCoherentUMA);
+    FEATURE_SUPPORT_GET_NODE_INDEXED(BOOL, m_dArchitecture1, IsolatedMMU);
+
+    // 18: D3D12 Options2
+    FEATURE_SUPPORT_GET(BOOL, m_dOptions2, DepthBoundsTestSupported);
+    FEATURE_SUPPORT_GET(D3D12_PROGRAMMABLE_SAMPLE_POSITIONS_TIER, m_dOptions2, ProgrammableSamplePositionsTier);
+
+    // 19: Shader Cache
+    FEATURE_SUPPORT_GET_NAME(D3D12_SHADER_CACHE_SUPPORT_FLAGS, m_dShaderCache, SupportFlags, ShaderCacheSupportFlags);
+
+private:
+    // Pointer to the underlying device
+    ID3D12Device* const m_pDevice;
+
+    // Stores the error code from initialization
+    HRESULT m_hStatus;
+
+    // Feature support data structs
+    D3D12_FEATURE_DATA_D3D12_OPTIONS m_dOptions;
+    std::vector<D3D12_FEATURE_DATA_ARCHITECTURE> m_dArchitecture; // Cat2 NodeIndex, should be deprecated

```

Comment:
I would like to avoid any dynamic allocation as well, but unfortunately the size of these data structures depend on the number of computing nodes at runtime. If we use statically allocated arrays (std::array) we must predetermine a size that's large enough to accommodate any possible node count, which can take up too much space for most of the time since most computers only have one or two of these graphics nodes.

However, if there's an (not too high) upper limit on the number of graphics nodes, or if there's a way to determine that number at compile time, I will vote for a static array.

Reply:
Forwarded to next PR.